parser.rst

```diff
@@ -0,0 +1,884 @@
+
+
+Consecuences or the ordered choice operator
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+.. _consecuences-of-ordered-choice:
+
+Although PEG may look like EBNF, but its meaning is quite different. The fact
+that in PEG parsers alternatives are ordered (which is the core of how PEG
+parsers work) has deep consecuences, other than removing ambiguity.
+
+If a rule has two alternatives and the first of them suceeds, the second one is **not** attempted
+even if the caller rule failes to parse the rest of the input. This is often described that the parser
+is "eager". To illustrate, this, consider the following two rules: ::
+
+    first_rule:  ( 'a' | 'aa' ) 'a'
+    second_rule: ('aa' | 'a'  ) 'a'
+
+In a relgular EBNF grammar, both rules specify the language ``{aa, aaa}`` but
+in PEG, one of these two rules accepts the string ``aaa`` but not the string
+``aa``. The other does the opposite -- it accepts the string the string ``aa``
+but not the string ``aaa``. The reason is that the rule ``('a'/'aa')'a'`` does
+not accept ``aaa`` because ``'a'/'aa'`` consumes the first ``a``, letting the
+final ``a`` in the rule consume the second, and leaving out the third ``a``.
+As the rule has suceeded, no attempt is ever made to go back and let
+``'a'/'aa'`` try the second alternative. The expression ``('aa'/'a')'a'`` does
+not accept ``aa`` because ``'aa'/'a'`` accepts all of ``aa``, leaving nothing
+for the final ``a``. Again, the second alternative of ``'aa'/'a'`` is not
+tried.
+
+.. caution::
+
+    The effects of ordered choice such as illustrated above, may be hidden by many levels of rules.
+
+For this reason, avoid writing rules where an alternative is contained in the next one,
+for example: ::
+
+    my_rule:
+        | 'if' expression 'then' block
+        | 'if' expression 'then' block 'else' block
+
+In this example, the second alternative will never be tried because the first one will
+suceed first (even if the input string has a ``'else' block`` that follows). To correctly
+write this rule you can simply alter the order: ::
+
+    my_rule:
+        | 'if' expression 'then' block
+        | 'if' expression 'then' block 'else' block
+
+In this case, if the input string doesn't have an ``'else' block``, the first alternative
+will fail and the second will be attempted without said part.
+
+Syntax
+------
+
+The grammar consists of a sequence of rules of the form: ::
+
+    rule_name: expression
+
+Optionally, a type can be included right after the rule name, which
+specifies the return type of the C or Python function corresponding to
+the rule: ::
+
+    rule_name[return_type]: expression
+
+If the return type is omitted, then a ``void *`` is returned in C and an
+``Any`` in Python.
+
+Grammar Expressions
+~~~~~~~~~~~~~~~~~~~
+
+``# comment``
+'''''''''''''
+
+Python-style comments.
+
+``e1 e2``
+'''''''''
+
+Match e1, then match e2.
+
+::
+
+    rule_name: first_rule second_rule
+
+.. _e1-e2-1:
```

Comment:
I don't understand this block.  It may be my fuzziness about .rst markup, but something seems extraneous or missing.  Same comment about other .. e1...(underscored) lines below.

Reply:
These are unused labels, I will remove them